### ISSUE TITLE
Support font collections

### ISSUE THREAD

**Issue by ZettaScript** [bevyengine/bevy#5773]
## What problem does this solve or what need does it fill?

OpenType font files may contain multiple fonts (e.g. regular, bold, italic, thin). When using such a file for displaying text with Bevy, only one of the available fonts can be used.

## What solution would you like?

There could be a way to specify which font of the collection must be used, maybe in `Font` or `TextStyle`.

## What alternative(s) have you considered?

Currently the only alternative is splitting the file into separate files for each font.

**Comment by Pietrek14** [comment]
I will try to implement that

**Comment by mockersf** [comment]
Bevy is currently using the [ab_glyph crate](https://crates.io/crates/ab_glyph) to read font data from files, it would probably need to handle those too if it doesn't yet (which from a quick look at their doc, they don't)

**Comment by Pietrek14** [comment]
Yeah, but can't we pass the index to the asset loader? Then we could use `FontVec::try_from_vec_and_index` to load the specific font from the collection,

**Comment by mockersf** [comment]
Ha, I missed they have that index parameter! that would make it possible but not simple to implement.

There are currently no way to pass additional information to the asset loader. I guess a possible solution would be to change the font loader to something like:
* load font index 0
* load font index 1
  * if it fails, returns a simple font
  * if not, load all fonts until one fails, save them as labelled subassets, and return a new asset type "font family" or something

**Comment by Pietrek14** [comment]
It seems that reading a font at a bigger index than available doesn't fail, but returns a valid font. That's unlucky.

I'll try to fix the issue in the ab_glyph itself, but it may take some time.

**Comment by Pietrek14** [comment]
@ZettaScript Could you share the font file? I'd like to use it for testing.

**Comment by ZettaScript** [comment]
https://txmn.tk/fonts/Cantarell-VF.otf

**Comment by Pietrek14** [comment]
Thanks a lot :)

**Comment by mockersf** [comment]
just checking, is it https://fonts.google.com/specimen/Cantarell available under the Open Font License?

**Comment by Pietrek14** [comment]
Are you sure that it's a font collection and not just a regular font? I think a collection should have a .ttc or .ttcf file extension, as per the [Wikipedia page](https://en.wikipedia.org/wiki/TrueType#TrueType_Collection).

**Comment by Waridley** [comment]
Wouldn't a simpler solution be to just use the label in the path? Then the user would need to use multiple `Font` handles for each font in the collection, but it's nearly trivial to implement.

```rust
let mut bytes = Vec::new();
reader.read_to_end(&mut bytes).await?;
let idx = ctx.asset_path()
    .label()
    .and_then(|lbl| lbl.parse()
        .inspect_err(|e| error!("Failed to parse label, using face index 0 instead: {e}"))
        .ok()
    )
    .unwrap_or(0);
Ok(Font::try_from_bytes_and_index(bytes, idx)?)
```

Just add the short function `try_from_bytes_and_index` to `Font`, and it's working.

Would this be a solution for your use case @ZettaScript, or would you need to load the entire family into one asset?

**Comment by eddyb** [comment]
Two potentially useful tools (that sadly require using `ttf-parser` itself, as `ab_glyph` intentionally only exposes a minimal subset of the full API):
- [`ttf_parser::fonts_in_collection`](https://docs.rs/ttf-parser/0.25.1/ttf_parser/fn.fonts_in_collection.html) can give a total count
- for each font (which hopefully isn't too expensive to "load just to peek at it"), [`.names()`](https://docs.rs/ttf-parser/0.25.1/ttf_parser/struct.Face.html#method.names) can be iterated to try to find a specific [`Name`](https://docs.rs/ttf-parser/0.25.1/ttf_parser/name/struct.Name.html) representative of the font
  - likely via one of the "known `name_id`s", e.g. [`ttf_parser::name_id::`](https://docs.rs/ttf-parser/0.25.1/ttf_parser/name_id/index.html)`FULL_NAME`
  - see also: https://learn.microsoft.com/en-us/typography/opentype/spec/name#name-ids

That way, the asset path may look like `my_font.ttf#MyFont Bold Italic`, though IIUC that doesn't mean there couldn't be other valid labels that find the same font (just `#Bold Italic` and/or allowing numerical indices etc.).

---

The other thing to keep in mind is variable fonts (which [`ab_glyph` thankfully exposes](https://docs.rs/ab_glyph/0.2.29/ab_glyph/trait.VariableFont.html)).

I was sad to learn that my favorite font doesn't use that for weights:
- https://github.com/be5invis/Iosevka/issues/300

But if a font does have the `wght` variation axis, instead of N `ttf` files (whether separately or bundled into a `ttc` files) - and N=9 is common, you can have a single file, and then set whatever weight is desired at runtime.

Speaking of which, [the `ab_glyph` docs for `VariableFont`] mention `Cantarell-VF.otf`.
Which is the same font that @ZettaScript brought up above.
The "VF" in the name means "variable font" I'm almost certain, and that means it *doesn't have to be* a collection (it could still be, just not for weight).

Variable fonts were brought up before in:
- https://github.com/bevyengine/bevy/issues/9725

That issue also mentions being able to register individual font files as being specific variations of a common font family - and *that* would also fit with collections (there's even some "WWS", i.e. "weight-width-slope" information in the actual font tables that might allow automating any necessary stylistic mapping, though I'm just skimming the OpenType spec and haven't poked at actual files).
